payroll_preparation_from_timesheet/models/account_analytic_line.py

```diff
@@ -0,0 +1,59 @@
+# © 2020 - today Numigi (tm) and all its contributors (https://bit.ly/numigiens)
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl).
+
+from odoo import api, fields, models, _
+from odoo.exceptions import ValidationError
+
+
+class AnalyticLine(models.Model):
+
+    _inherit = 'account.analytic.line'
+
+    payroll_preparation_line_ids = fields.One2many(
+        'payroll.preparation.line',
+        'timesheet_id',
+        'Payroll Entries',
+    )
+
+    @api.multi
+    def unlink(self):
+        for line in self:
+            if line._has_payroll_entries():
+                raise ValidationError(_(
+                    "You may not delete the timesheet line {} "
+                    "because it is linked to a payroll entry."
+                ).format(line.display_name))
+        return super().unlink()
+
+    @api.multi
+    def write(self, vals):
+        is_super_admin = self.env.user._is_superuser()
+        if not is_super_admin:
+            self._check_protected_vals_for_payroll_entries(vals)
+        return super().write(vals)
+
+    def _check_protected_vals_for_payroll_entries(self, vals):
+        protected_fields = self._get_payroll_entry_protected_fields()
+        protected_field_modified = protected_fields.intersection(vals)
+        for line in self:
+            if protected_field_modified and line._has_payroll_entries():
+                raise ValidationError(_(
+                    "You may not modify the timesheet line {} "
+                    "because it is linked to a payroll entry."
+                ).format(line.display_name))
+
+    def _get_payroll_entry_protected_fields(self):
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Method could be a function](https://app.codacy.com/manual/numigi/odoo-timesheet-addons/pullRequest?prid=4859503)